tinygrad/ops.py

```diff
@@ -238,54 +238,49 @@ def backward(ctx, grad_output):
 
 # ************* pooling ops *************
 
-def stack_for_pool(x, py, px):
-  my, mx = (x.shape[2]//py)*py, (x.shape[3]//px)*px
-  stack = []
-  xup = x[:, :, :my, :mx]
-  for Y in range(py):
-    for X in range(px):
-      stack.append(xup[:, :, Y::py, X::px][None])
-  return np.concatenate(stack, axis=0)
-
-def unstack_for_pool(fxn, s, py, px):
+def pool(x, py, px, fxn=None):
+  s = x if fxn else x.shape
   my, mx = (s[2]//py)*py, (s[3]//px)*px
-  for Y in range(py):
-    for X in range(px):
+  stack, xup = (None,None) if fxn else (list([]),x[:,:,:my,:mx])
+  for Y, X in itertools.product(range(py), range(px)):
+    if fxn:
       ll = fxn(Y*px+X)
       if X == 0 and Y == 0:
         ret = np.zeros(s, dtype=ll.dtype)
       ret[:, :, Y:my:py, X:mx:px] = ll
-  return ret
+    else:
+      stack.append(xup[:, :, Y::py, X::px][None])
+  return ret if fxn else np.concatenate(stack, axis=0)
 
 class MaxPool2D(Function):
   @staticmethod
   def forward(ctx, x, kernel_size=(2, 2)):
-    stack = stack_for_pool(x, *kernel_size)
+    stack = pool(x, *kernel_size)
     idxs = np.argmax(stack, axis=0)
     ctx.save_for_backward(idxs, x.shape)
     return np.max(stack, axis=0)
 
   @staticmethod
   def backward(ctx, grad_output):
     idxs,s = ctx.saved_tensors
-    return unstack_for_pool(
-      lambda idx: grad_output * (idxs == idx),
-      s, *ctx.kernel_size)
+    return pool(s, *ctx.kernel_size,
+      fxn = lambda idx: grad_output * (idxs == idx)
+      )
```

Comment:
This paren can go on the above line.

Reply:
idk what u mean by this.

Reply:
fxn = lambda idx: grad_output * (idxs == idx))